File: backend/app/schemas/analytics.py
from datetime import date
from typing import List, Optional
from pydantic import BaseModel, Field


class TimeSeriesDataPoint(BaseModel):
    """A single data point in a time series."""
    period: str  # e.g., "2024-01", "2024-W05", "2024-01-15"
    count: int


class CasesOverTimeResponse(BaseModel):
    """Cases over time for line/area charts."""
    data: List[TimeSeriesDataPoint]
    period_type: str  # "day", "week", "month"


class EntityDistributionItem(BaseModel):
    """Entity type distribution item."""
    entity_type: str
    count: int
    percentage: float = Field(ge=0, le=100)


class EntityDistributionResponse(BaseModel):
    """Entity distribution by type for pie charts."""
    data: List[EntityDistributionItem]
    total: int


class TopEntityItem(BaseModel):
    """A top entity item."""
    name: str
    entity_type: str
    case_count: int
    occurrence_count: int


class TopEntitiesResponse(BaseModel):
    """Top entities by case appearances."""
    data: List[TopEntityItem]


class CourtDistributionItem(BaseModel):
    """Cases distribution by court."""
    court_name: str
    count: int
    percentage: float = Field(ge=0, le=100)


class CourtDistributionResponse(BaseModel):
    """Cases by court for bar charts."""
    data: List[CourtDistributionItem]
    total: int


class StatusDistributionItem(BaseModel):
    """Cases by processing status."""
    status: str
    count: int


class AnalyticsSummaryResponse(BaseModel):
    """Overall analytics summary."""
    total_cases: int
    total_entities: int
    unique_entities: int
    avg_entities_per_case: float
    cases_this_month: int
    entities_this_month: int
    status_distribution: List[StatusDistributionItem]


class TrendsResponse(BaseModel):
    """Combined trends data for dashboard."""
    cases_over_time: List[TimeSeriesDataPoint]
    entity_distribution: List[EntityDistributionItem]
    top_entities: List[TopEntityItem]
    court_distribution: List[CourtDistributionItem]

File: frontend/src/services/analytics.ts
import api from './api';

// Types for analytics data

export interface TimeSeriesDataPoint {
  period: string;
  count: number;
}

export interface CasesOverTimeResponse {
  data: TimeSeriesDataPoint[];
  period_type: string;
}

export interface EntityDistributionItem {
  entity_type: string;
  count: number;
  percentage: number;
}

export interface EntityDistributionResponse {
  data: EntityDistributionItem[];
  total: number;
}

export interface TopEntityItem {
  name: string;
  entity_type: string;
  case_count: number;
  occurrence_count: number;
}

export interface TopEntitiesResponse {
  data: TopEntityItem[];
}

export interface CourtDistributionItem {
  court_name: string;
  count: number;
  percentage: number;
}

export interface CourtDistributionResponse {
  data: CourtDistributionItem[];
  total: number;
}

export interface StatusDistributionItem {
  status: string;
  count: number;
}

export interface AnalyticsSummaryResponse {
  total_cases: number;
  total_entities: number;
  unique_entities: number;
  avg_entities_per_case: number;
  cases_this_month: number;
  entities_this_month: number;
  status_distribution: StatusDistributionItem[];
}

export interface TrendsResponse {
  cases_over_time: TimeSeriesDataPoint[];
  entity_distribution: EntityDistributionItem[];
  top_entities: TopEntityItem[];
  court_distribution: CourtDistributionItem[];
}

export const analyticsService = {
  async getSummary(): Promise<AnalyticsSummaryResponse> {
    const response = await api.get<AnalyticsSummaryResponse>('/api/analytics/summary');
    return response.data;
  },

  async getCasesOverTime(
    period: 'day' | 'week' | 'month' = 'month',
    months: number = 12
  ): Promise<CasesOverTimeResponse> {
    const response = await api.get<CasesOverTimeResponse>('/api/analytics/cases-over-time', {
      params: { period, months }
    });
    return response.data;
  },

  async getEntityDistribution(): Promise<EntityDistributionResponse> {
    const response = await api.get<EntityDistributionResponse>('/api/analytics/entity-distribution');
    return response.data;
  },

  async getTopEntities(limit: number = 10, entityType?: string): Promise<TopEntitiesResponse> {
    const response = await api.get<TopEntitiesResponse>('/api/analytics/top-entities', {
      params: { limit, entity_type: entityType }
    });
    return response.data;
  },

  async getCourtDistribution(): Promise<CourtDistributionResponse> {
    const response = await api.get<CourtDistributionResponse>('/api/analytics/courts');
    return response.data;
  },

  async getTrends(
    period: 'day' | 'week' | 'month' = 'month',
    months: number = 6
  ): Promise<TrendsResponse> {
    const response = await api.get<TrendsResponse>('/api/analytics/trends', {
      params: { period, months }
    });
    return response.data;
  },
};
